### PR TITLE
Features/asset retargeting intermediatedlls

### PR DESCRIPTION
**Pull request by gejohnst**
Overview
---

Changes
---
- Fixes: # .

### CODE REVIEW

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 1]
```diff
@@ -217,10 +219,14 @@ private static async Task ProcessYamlFile(string filePath, string targetPath, Di
                             {
                                 line = $"  m_Script: {{fileID: {tuple.Item2}, guid: {tuple.Item1}, type: 3}}";
                             }
+                            else if (nonClassDictionary.Contains(filePath))
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: This is supposed to be the guid, not the yaml file path, right? #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 2]
```diff
@@ -250,6 +256,8 @@ private static bool IsYamlFile(string filePath)
             }
         }
 
+        private static List<string> nonClassDictionary = new List<string>();
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: could you put this above along with the rest of dictionaries?

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 3]
```diff
@@ -328,6 +337,11 @@ private static bool IsYamlFile(string filePath)
                             if (!(monoScript is null) && AssetDatabase.TryGetGUIDAndLocalFileIdentifier(monoScript, out string guid, out long fileId))
                             {
                                 Type type = monoScript.GetClass();
+
+                                if (type.Namespace == null || !type.Namespace.Contains("Microsoft.MixedReality.Toolkit"))
+                                {
+                                    Debug.LogErrorFormat("Type {0} is not a member of the Microsoft.MixedReality.Toolkit namespace", type.Name);
+                                }
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs:351]: We should throw an exception here, as this is a hard stop; we just hit something that is a bug in the MRTK codebase
 #Resolved

Comment (gejohnst) [reply Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs:351]: There are currently 9 of these, btw - all of them have a null namespace.  should that be a separate case? #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs (hunk 2 above)
Comment (gejohnst) [reply Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: will do

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 4]
```diff
@@ -343,10 +357,159 @@ private static bool IsYamlFile(string filePath)
...
+                {
+                    case "Win":
+                        subfolderName = "Standalone";
+                        break;
+                    case "WindowsStoreApps":
+                        subfolderName = "UAP";
+                        break;
+                    default:
+                        continue;
+                }
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs:387]: We should be more explicit, perhaps declare a dictionary that maps between folder we expect to exist and it's output location. And then if the folder doesn't exist, we again thrown an error, as we can't create a complete package without it #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs (hunk 1 above)
Comment (gejohnst) [reply Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: yeah, that makes sense #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 5]
```diff
@@ -343,10 +357,159 @@ private static bool IsYamlFile(string filePath)
...
+        {
+            int lengthOfPrefix = Application.dataPath.IndexOf("Assets");
+            Dictionary<string, string> dllGuids = new Dictionary<string, string>();
+            string[] asmdefFiles = Directory.GetFiles(Application.dataPath, "*.asmdef", SearchOption.AllDirectories);
+
+            foreach (string asmdefFile in asmdefFiles)
+            {
+                string asmdefText = AssetDatabase.LoadAssetAtPath<AssemblyDefinitionAsset>(asmdefFile.Substring(lengthOfPrefix)).text;
+                string dllName = JsonUtility.FromJson<AssemblyDefinitionStub>(asmdefText).name;
+                string guid = File.ReadAllLines(string.Format("{0}.meta", asmdefFile))[1].Substring(6);
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: >string.Format("{0}.meta", asmdefFile) [](start = 48, length = 37)

could us the $"{asmdefFile}.meta" syntax. #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs (hunk 4 above)
Comment (gejohnst) [reply Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs:387]: Sounds good - I'll get that in and submit the update with the rest of the changes #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 6]
```diff
@@ -217,10 +228,14 @@ private static async Task ProcessYamlFile(string filePath, string targetPath, Di
                             {
                                 line = $"  m_Script: {{fileID: {tuple.Item2}, guid: {tuple.Item1}, type: 3}}";
                             }
+                            else if (nonClassDictionary.ContainsKey(guid))
+                            {
+                                Debug.LogErrorFormat("A script without a class ({0}) is being processed.", nonClassDictionary[guid]);
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: I wonder if we want to throw exception here, cause we are in a weird state then #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 7]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+
+
+
+        private static void CopyIntermediaryAssemblies(string outputPath)
+        {
+            List<DirectoryInfo> outputDirectories = new List<DirectoryInfo>();
+            string playerDataCachePath = Application.dataPath.Replace("Assets", "Library/PlayerDataCache");
+            DirectoryInfo playerDataCahce = new DirectoryInfo(playerDataCachePath);
+            DirectoryInfo[] dllStores = playerDataCahce.GetDirectories("*", SearchOption.TopDirectoryOnly);
+            string subfolderName = string.Empty;
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: You could move the variable declaration to be inline at 392 (save a line)

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 8]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+        }
+
+
+
+        private static void CopyIntermediaryAssemblies(string outputPath)
+        {
+            List<DirectoryInfo> outputDirectories = new List<DirectoryInfo>();
+            string playerDataCachePath = Application.dataPath.Replace("Assets", "Library/PlayerDataCache");
+            DirectoryInfo playerDataCahce = new DirectoryInfo(playerDataCachePath);
+            DirectoryInfo[] dllStores = playerDataCahce.GetDirectories("*", SearchOption.TopDirectoryOnly);
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: >dllStores [](start = 28, length = 9)

this is no longer used

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 9]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+            string subfolderName = string.Empty;
+            foreach(KeyValuePair<string, string> sourceToOutputPair in sourceToOutputFolders)
+            {
+                DirectoryInfo directory = new DirectoryInfo(Application.dataPath.Replace("Assets", sourceToOutputPair.Key));
+                if (!directory.Exists)
+                {
+                    throw new InvalidDataException($"The source directory {sourceToOutputPair.Key} does not exist.");
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: >The source directory [](start = 53, length = 20)

Phrase it as: "The required platform intermediary build directory {sourceToOutputPair.Key} does not exist. Was the build succesful?" #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 10]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+                }
+
+                subfolderName = sourceToOutputPair.Value;
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: >sourceToOutputPair.Value [](start = 32, length = 24)

Could get rid of subFolderName variable all together

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 11]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+                guid = CycleGuidForward(guid);
+                dllGuids.Add($"{dllName}.dll", guid);
+            }
+
+            //Load the sample meta files
+            DirectoryInfo assetDirectory = new DirectoryInfo(Application.dataPath);
+            FileInfo editorMetaFile = assetDirectory.GetFiles("*sampleEditorDllMeta.txt", SearchOption.AllDirectories).FirstOrDefault();
+            if (editorMetaFile == null)
+            {
+                Debug.LogError("Could not find sample editor dll.meta file");
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: > Debug.LogError("Could not find sample editor dll.meta file"); [](start = 15, length = 62)

We should throw an error here #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 12]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+                        dllGuid = dllGuids[file.Name];
+                        dllGuid = CycleGuidForward(dllGuid);
+                        if (isUAP)
+                        {
+                            dllGuid = CycleGuidForward(dllGuid);
+                        }
+
+                        if (dllGuids.ContainsKey(file.Name))
+                        {
+                            metaFilePath = string.Format("{0}.meta", file.FullName);
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: > string.Format("{0}.meta", file.FullName); [](start = 42, length = 42)

Could use the new syntax and also inline the declaration of the field (remove from line 476 #Resolved

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 13]
```diff
@@ -343,10 +364,164 @@ private static bool IsYamlFile(string filePath)
...
+                        break;
+                    case '9':
+                        guidBuilder.Append('a');
+                        break;
+                    default:
+                        guidBuilder.Append((char)(hexit + 1));
+                        break;
+                }
+            }
+            return guidBuilder.ToString();
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs:519]: You could literally just cycle the first character :)

File: Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs [hunk 14]
```diff
@@ -397,5 +572,10 @@ private static void RecursiveFolderCleanup(DirectoryInfo folder)
             }
         }
     }
+
+    struct AssemblyDefinitionStub
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs]: >struct [](start = 4, length = 6)

nit: put private

**Review by andreiborodin — APPROVED** [review]
:shipit:

**Pull request merged**